openpgp/key_generation.go

```diff
@@ -29,150 +28,138 @@ func NewEntity(name, comment, email string, config *packet.Config) (*Entity, err
-
-	} else {
-		return nil, errors.InvalidArgumentError("unsupported public key algorithm")
+	// Generate a primary signing key
+	primaryPrivRaw, err := newSigner(config)
+	if err != nil {
+		return nil, err
 	}
+	primary := packet.NewSignerPrivateKey(creationTime, primaryPrivRaw)
 
-	e := &Entity{
-		PrimaryKey: pubPrimary,
-		PrivateKey: privPrimary,
-		Identities: make(map[string]*Identity),
-	}
 	isPrimaryId := true
-	e.Identities[uid.Id] = &Identity{
-		Name:   uid.Id,
-		UserId: uid,
-		SelfSignature: &packet.Signature{
-			CreationTime: creationTime,
-			SigType:      packet.SigTypePositiveCert,
-			PubKeyAlgo:   primarykeyAlgorithm,
-			Hash:         config.Hash(),
-			IsPrimaryId:  &isPrimaryId,
-			FlagsValid:   true,
-			FlagSign:     true,
-			FlagCertify:  true,
-			// Set MDC true by default, see 5.8 vs. 5.14
-			MDC:          true,
-			AEAD:         config.AEAD() != nil,
-			IssuerKeyId:  &e.PrimaryKey.KeyId,
-		},
+	selfSignature := &packet.Signature{
+		SigType:            packet.SigTypePositiveCert,
+		PubKeyAlgo:         config.PublicKeyAlgorithm(),
+		Hash:               config.Hash(),
+		CreationTime:       creationTime,
+		IssuerKeyId:        &primary.PublicKey.KeyId,
+		IsPrimaryId:        &isPrimaryId,
+		FlagsValid:         true,
+		FlagSign:           true,
+		FlagCertify:        true,
+		MDC:                true, // true by default, see 5.8 vs. 5.14
+		AEAD:               config.AEAD() != nil,
 	}
-	e.Identities[uid.Id].Signatures = append(e.Identities[uid.Id].Signatures, e.Identities[uid.Id].SelfSignature)
 
-	// If the user passes in a DefaultHash via packet.Config,
-	// set the PreferredHash for the SelfSignature.
-	if config != nil && config.DefaultHash != 0 {
-		e.Identities[uid.Id].SelfSignature.PreferredHash = []uint8{hashToHashId(config.DefaultHash)}
+	// Set the PreferredHash for the SelfSignature from the packet.Config.
+	// If it is not the must-implement algorithm from rfc4880bis, append that.
+	selfSignature.PreferredHash = []uint8{hashToHashId(config.Hash())}
+	if (config.Hash() != crypto.SHA256) {
+		selfSignature.PreferredHash = append(selfSignature.PreferredHash, hashToHashId(crypto.SHA256))
 	}
 
 	// Likewise for DefaultCipher.
-	if config != nil && config.DefaultCipher != 0 {
-		e.Identities[uid.Id].SelfSignature.PreferredSymmetric = []uint8{uint8(config.DefaultCipher)}
+	selfSignature.PreferredSymmetric = []uint8{uint8(config.Cipher())}
+	if (config.Cipher() != packet.CipherAES128) {
+		selfSignature.PreferredSymmetric = append(selfSignature.PreferredSymmetric, uint8(packet.CipherAES128))
 	}
 
 	// And for DefaultMode.
-	if config.AEAD() != nil && config.AEAD().DefaultMode != 0 {
-		e.Identities[uid.Id].SelfSignature.PreferredAEAD = []uint8{uint8(config.AEAD().DefaultMode)}
+	selfSignature.PreferredAEAD = []uint8{uint8(config.AEAD().Mode())}
+	if (config.AEAD().Mode() != packet.AEADModeEAX) {
+		selfSignature.PreferredAEAD = append(selfSignature.PreferredAEAD, uint8(packet.AEADModeEAX))
```

Comment:
Yeah, I understand, and sorry for reverting that part. But https://github.com/ProtonMail/crypto/pull/50/commits/686f7e1fe233b131e895c2be72c4b16c923170e1 made that code smaller by half, an IMHO less repetitive.

> Algorithm preferences are a `config` property and in my opinion should be managed there for modularity

Arguably, what the code is doing is combining the preferences of the user (e.g. AES256) with the defaults of the spec (e.g. AES128). IMO it's consistent to have a function returning the former in `config` and code doing the latter in `openpgp`.

Reply:
> But 686f7e1 made that code smaller by half
Because you also removed explicit sanitization. It will be done implicitly eventually, but again, not good for modularity in my opinion.

Reply:
Yeah. We can add it back in `Cipher()` and `Hash()` in the later PR where we disallow 3DES and CAST5, with something like
```patch
 func (c *Config) Cipher() CipherFunction {
-       if c == nil || uint8(c.DefaultCipher) == 0 {
+       if c == nil || uint8(c.DefaultCipher) == 0 || (
+               c.DefaultCipher != CipherAES256 &&
+               c.DefaultCipher != CipherAES192) {
                return CipherAES128
        }
        return c.DefaultCipher
 }
```
The way it was, was a bit inconsistent between `Cipher()` and `PreferredSymmetricAlgorithms()`, where the former could still return a different algorithm. So I think it's better to do it in `Cipher()`, and after we disallow 3DES and CAST5, it should still be pretty short.

Reply:
I fail to see the inconsistency: `PreferredSymmetricAlgorithms()` returns a list consisting of the specified algorithm and the must-implement algorithm. `Cipher()` returns the specified algorithm.

Reply:
Yeah, but you were doing this validation in `PreferredSymmetricAlgorithms()`, but not in `Cipher()`

Reply:
I still don't see how that makes it inconsistent; `Cipher()` assumes that `DefaultCipher` is set from the list imported into the `openpgp` package (for `Hash()`, another story; it should not be importing from `crypto`), and that makes the "validation" completely equivalent to the one in `PreferredCipher`. If you wanted you could have changed only the validation instead of removing the functions altogether (but I prefer having an explicit validation anyway for better control).

Reply:
> I still don't see how that makes it inconsistent; Cipher() assumes that DefaultCipher is set from the list imported into the openpgp package

But the same argument applies to `PreferredSymmetricAlgorithms`, no? Why is it important to explicitly validate `c.DefaultCipher` in `PreferredSymmetricAlgorithms`, if it's not important in `Cipher()`?

Reply:
I think it is important too in `Cipher()` and company, of course, but we were not using this in key generation.

Reply:
Are these function only called when generating a new entity?

Reply:
Also when serializing (see ` openpgp/packet/symmetric_key_encrypted.go`, `openpgp/packet/aead_encrypted.go`).

Reply:
In gopenpgp we use the serializing functions.
On the other hand an existing key should not be altered when serialized, imho, no matter the sanitation

Reply:
No, we also use `Cipher()` when symmetrically encrypting a message, for example

Reply:
`Cipher()` is only called when encrypting a message (and generating a key). We don't have support for serializing an existing encrypted message. It's also not called when encrypting an existing key.

The only issue could arise when using `SerializeSymmetricKeyEncryptedReuseKey`, which is used when you have an existing session key (data + algorithm). In my opinion, this function should be modified to take cipher algorithms directly, instead of a `config`, since we're not really expressing a preference in that case, but rather reusing existing data. Since we added this function ourselves, we could probably change it.